src/main/java/pl/most/typer/service/footballservice/matches/MatchesServiceDefault.java

```diff
@@ -90,32 +83,72 @@ public HttpStatus getMatchInfoFromExternalApi(Integer competitionId) {
     }
 
     @Override
-    public void saveAll(List<Match> matches) {
+    public void saveAll(List<Match> matches) throws BadResourceException, ResourceAlreadyExistsException {
         for (Match match : matches) {
-            Optional<Match> byApiId = matchesRepository.findByApiId(match.getApiId());
-            byApiId.orElseGet(() -> matchesRepository.save(match));
+            if (!StringUtils.isEmpty(match)) {
+                if (match.getApiId() != null && existsByApiId(match.getApiId())) {
+                    ResourceAlreadyExistsException ex = new ResourceAlreadyExistsException("Match with id: " +
+                            match.getApiId() + " already exists");
+                    ex.setResource("Match");
+                    ex.setIssue("id");
+                } else {
+                    matchesRepository.save(match);
+                }
+            } else {
+                BadResourceException ex = new BadResourceException("Failed to save Match");
+                ex.setErrorMessage("Score is null or empty");
+            }
+
         }
     }
 
+    @Override
+    public Match findByApiId(Integer apiId) throws ResourceNotFoundException {
+        Optional<Match> matchOptional = matchesRepository.findByApiId(apiId);
+        return matchOptional.orElseThrow(() -> {
+            ResourceNotFoundException ex = new ResourceNotFoundException("Cannot find Match with id: " + apiId);
+            ex.setResource("match");
+            return ex;
+        });
+    }
+
     @Override
     public List<Match> findAllByCompetition(Competition competition) {
-        return matchesRepository.findAllByCompetitionOrderByUtcDateDesc(competition);
+        Optional<List<Match>> optionalMatches = Optional.ofNullable(matchesRepository.findAllByCompetitionOrderByUtcDateDesc(competition));
```

Comment:
W sumie to nie wiem czy to nie jest przekombinowane z wstawianiem wyniku do optionala, po to zeby puscic sprobowac rzucic wyjatek. Mozna zrobic zwyklego if matches.isEmpty() -> throw Exception....